Source/Core/Core/AchievementManager.cpp

```diff
@@ -1016,694 +520,461 @@ void AchievementManager::FilereaderClose(void* file_handle)
   delete static_cast<FilereaderState*>(file_handle);
 }
 
-AchievementManager::ResponseType AchievementManager::VerifyCredentials(const std::string& password)
+void AchievementManager::LoginCallback(int result, const char* error_message, rc_client_t* client,
+                                       void* userdata)
 {
-  rc_api_login_response_t login_data{};
-  std::string username, api_token;
+  if (result != RC_OK)
   {
-    std::lock_guard lg{m_lock};
-    username = Config::Get(Config::RA_USERNAME);
-    api_token = Config::Get(Config::RA_API_TOKEN);
+    WARN_LOG_FMT(ACHIEVEMENTS, "Failed to login {} to RetroAchievements server.",
+                 Config::Get(Config::RA_USERNAME));
+    return;
   }
-  rc_api_login_request_t login_request = {
-      .username = username.c_str(), .api_token = api_token.c_str(), .password = password.c_str()};
-  ResponseType r_type = Request<rc_api_login_request_t, rc_api_login_response_t>(
-      login_request, &login_data, rc_api_init_login_request, rc_api_process_login_response);
-  if (r_type == ResponseType::SUCCESS)
+
+  const rc_client_user_t* user;
   {
-    INFO_LOG_FMT(ACHIEVEMENTS, "Successfully logged in {} to RetroAchievements server.", username);
-    std::lock_guard lg{m_lock};
-    if (username != Config::Get(Config::RA_USERNAME))
-    {
-      INFO_LOG_FMT(ACHIEVEMENTS, "Attempted to login prior user {}; current user is {}.", username,
-                   Config::Get(Config::RA_USERNAME));
-      Config::SetBaseOrCurrent(Config::RA_API_TOKEN, "");
-      return ResponseType::EXPIRED_CONTEXT;
-    }
-    Config::SetBaseOrCurrent(Config::RA_API_TOKEN, login_data.api_token);
-    m_display_name = login_data.display_name;
-    m_player_score = login_data.score;
+    std::lock_guard lg{AchievementManager::GetInstance().GetLock()};
+    user = rc_client_get_user_info(client);
   }
-  else
+  std::string config_username = Config::Get(Config::RA_USERNAME);
+  if (config_username.compare(user->username) != 0)
   {
-    WARN_LOG_FMT(ACHIEVEMENTS, "Failed to login {} to RetroAchievements server.", username);
+    bool case_mismatch = (config_username.length() == strlen(user->username));
+    if (case_mismatch)
+      for (int ix = 0; ix < config_username.length(); ix++)
```

Comment:
This loop index should be a `std::size_t`. It currently produces a `-Wsign-compare` warning and has no need to be an int.

Reply:
Also the bounds-checked `std::string::at` is superfluous when iterating over the string you got the length from. Better yet, `Common::CaseInsensitiveEquals` seems like what you really should put here.